Runtime/Core/ConfinerOven.cs

```diff
@@ -0,0 +1,307 @@
+                    {
+                        shrinking = true;
+                        break;
+                    }
+                }
+            }
+        }
+     
+        /// <summary>
+        /// Converts and returns shrinkable polygons from a polygons
+        /// </summary>
+        public static List<List<ShrinkablePolygon>> CreateShrinkablePolygons(
+            in List<List<Vector2>> paths, in float aspectRatio)
+        {
+            int numPaths = paths == null ? 0 : paths.Count;
+            var shrinkablePolygons = new List<List<ShrinkablePolygon>>(numPaths);
+            if (numPaths > 0)
+            {
+                float minX = float.MaxValue, maxX = float.MinValue;
+                float minY = float.MaxValue, maxY = float.MinValue;
+                for (int i = 0; i < numPaths; ++i)
+                {
+                    var newShrinkablePolygon = new ShrinkablePolygon(paths[i], aspectRatio);
+                    int numPoints = newShrinkablePolygon.m_Points.Count;
+                    for (int j = 0; j < numPoints; ++j)
+                    {
+                        var p = newShrinkablePolygon.m_Points[j].m_Position;
+                        minX = Mathf.Min(minX, p.x);
+                        minY = Mathf.Min(minY, p.y);
+                        maxX = Mathf.Max(maxX, p.x);
+                        maxY = Mathf.Max(maxY, p.y);
+                    }
+                    shrinkablePolygons.Add(new List<ShrinkablePolygon> { newShrinkablePolygon });
+                }
+
+                float squareSize = Mathf.Min(maxX - minX, maxY - minY);
+                for (int i = 0; i < shrinkablePolygons.Count; ++i)
+                {
+                    shrinkablePolygons[i][0].m_MinArea = squareSize / 100f;
+                }
+            }
+            return shrinkablePolygons;
+        }
+        
+        /// <summary>
+        /// Converts and returns a prebaked ConfinerState for the input frustumHeight.
+        /// </summary>
+        public ConfinerState GetConfinerAtFrustumHeight(float frustumHeight)
+        {
+            ConfinerState result = new ConfinerState();
+            for (int i = m_confinerStates.Count - 1; i >= 0; --i)
+            {
+                if (m_confinerStates[i].m_FrustumHeight <= frustumHeight)
+                {
+                    if (i == m_confinerStates.Count - 1)
+                    {
+                        result = m_confinerStates[i];
+                    }
+                    else if (Math.Abs(m_confinerStates[i].m_State - m_confinerStates[i + 1].m_State) < 1e-6f)
+                    {
+                        // blend between m_confinerStates with same m_State
+                        result = ConfinerStateLerp(m_confinerStates[i], m_confinerStates[i+1], frustumHeight);
+                    }
+                    else
+                    {
+                        // choose m_confinerStates with windowSize closer to frustumHeight
+                        result = 
```

Comment:
I think this is a cause of the bad results when step size is large.  In this case, for a whole step size around the state change, you are arbitrarily choosing something that's guaranteed to be wrong because you're snapping to a boundary.  That error is not noticeable if the step size is very small.

I think a fix would involve precisely locating the boundary points over which the ConfinerState is non-lerpable (perhaps by initially having a large step size, and using a binary search when the state changes, with increasingly smaller step sizes, to locate the state transition point precisely).  

You would then make generate a ConfinerState a very small distance on either side of the state change boundary, and keep only those ConfinerStates.

Reply:
Yes, this is a bit tricky. I have to implement it at bake state.